### PR TITLE
(fixes #211) RedshiftUnloadActivity now works when it contains expres…

### LINKED ISSUE #211: RedhishiftUnloadActivity fail when containing expressions with `'`.
RedshiftUnloadActivity is replacing all `'` in the script to `\\'`, this works when the script does not contain expressions that have `'` in them. However, when the expression contains `'` (e.g. `#{format(@scheduledStartTime,'yyyy-MM-dd')}`) it will also replace the `'` in the expression as well. This is failing the creation of the pipeline with a `Cannot evaluate expression` error.

### PR DESCRIPTION
**Pull request by realstraw**
…sions

### CODE REVIEW

File: contrib/activity/definition/src/main/scala/com/krux/hyperion/activity/RedshiftUnloadActivity.scala [hunk 1]
```diff
@@ -37,8 +40,42 @@ case class RedshiftUnloadActivity private (
...
+          prepareScript(
+            expTail, expLevel, if (curChar == '\'') result ++= "\\\\'" else result + curChar)
+      } else if (expLevel == -1){  // the previous char is '#'
+        if (curChar == '{')  // start of an expression
+          prepareScript(expTail, 1, result + curChar)
+        else
+          prepareScript(
+            expTail, 0, if (curChar == '\'') result ++= "\\\\'" else result + curChar)
+      } else {  // inside a expression block
+        if (curChar == '}')
```
Comment (sethyates) [inline contrib/activity/definition/src/main/scala/com/krux/hyperion/activity/RedshiftUnloadActivity.scala]: I think this is unsafe approach as you could conceivably have an unmatched `{` or `}` in a string inside of an expression.

Comment (realstraw) [inline contrib/activity/definition/src/main/scala/com/krux/hyperion/activity/RedshiftUnloadActivity.scala]: Yes, I'll fix it

**Comment by realstraw** [comment]
@sethyates above commit handles the "{" and '{' case in expression

File: contrib/activity/definition/src/test/scala/com/krux/hyperion/activity/RedshiftUnloadActivitySpec.scala [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+      val `*password` = "mockpass"
+      val databaseName = "mock_db"
+    }
+
+    "Produce the correct unload script" in {
+      val testingQuery = """
+        |select * from t where
+        |id = 'myid'
+        |and {tim'e} = #{format(@actualRunTime, 'yyyy-MM-dd')}
+        |and someOtherWierdFormat = #{"{ } a'dfa {" + ' { ex"aef { }'}
```
Comment (sethyates) [inline contrib/activity/definition/src/test/scala/com/krux/hyperion/activity/RedshiftUnloadActivitySpec.scala]: Tyop

File: contrib/activity/definition/src/main/scala/com/krux/hyperion/activity/RedshiftUnloadActivity.scala [hunk 3]
```diff
@@ -37,13 +40,86 @@ case class RedshiftUnloadActivity private (
...
 
+  /**
+   * Given the start of exp, seek the end of expression returning the expression block and the rest
+   * of the string. Note that expression is not a nested structure and the only legitimate '{' or
+   * '}' within a expression is within quotes (i.e. '"' or "'")
+   *
+   * @note this does not handle the case that expression have escaped quotes (i.e. "\"" or '\'')
+   */
+  @tailrec
+  private def seekEoe(
```
Comment (sethyates) [inline contrib/activity/definition/src/main/scala/com/krux/hyperion/activity/RedshiftUnloadActivity.scala]: `Eoe` is a non-obvious name.  Would be better to be more explicit e.g., `seekEndOfInput`

File: contrib/activity/definition/src/main/scala/com/krux/hyperion/activity/RedshiftUnloadActivity.scala [hunk 4]
```diff
@@ -37,13 +40,86 @@ case class RedshiftUnloadActivity private (
...
+      exp: String,
+      quote: Option[Char] = None,
+      expPart: StringBuilder = StringBuilder.newBuilder
+    ): (String, String) = {
+
+    if (exp == "") {
```
Comment (sethyates) [inline contrib/activity/definition/src/main/scala/com/krux/hyperion/activity/RedshiftUnloadActivity.scala]: `.isEmpty`

File: contrib/activity/definition/src/main/scala/com/krux/hyperion/activity/RedshiftUnloadActivity.scala [hunk 5]
```diff
@@ -37,13 +40,86 @@ case class RedshiftUnloadActivity private (
...
+      throw new RuntimeException("expression started but not ended")
+    } else {
+      val curChar = exp.head
+      val next = exp.tail
+
+      quote match {
+        case Some('\'') =>  // if started in single quoted string
+          if (curChar == '\'') seekEoe(next, None, expPart + curChar)
+          else seekEoe(next, quote, expPart + curChar)
+        case Some('"') =>
```
Comment (sethyates) [inline contrib/activity/definition/src/main/scala/com/krux/hyperion/activity/RedshiftUnloadActivity.scala]: Can't you combine this and the previous as follows?
````
case Some(quoteChar) =>  // if started in quoted string
          if (curChar == quoteChar) seekEoe(next, None, expPart + curChar)
          else seekEoe(next, quote, expPart + curChar)
````

File: contrib/activity/definition/src/main/scala/com/krux/hyperion/activity/RedshiftUnloadActivity.scala [hunk 6]
```diff
@@ -37,13 +40,86 @@ case class RedshiftUnloadActivity private (
...
+
+    if (exp == "") {
+      throw new RuntimeException("expression started but not ended")
+    } else {
+      val curChar = exp.head
+      val next = exp.tail
+
+      quote match {
+        case Some('\'') =>  // if started in single quoted string
+          if (curChar == '\'') seekEoe(next, None, expPart + curChar)
```
Comment (sethyates) [inline contrib/activity/definition/src/main/scala/com/krux/hyperion/activity/RedshiftUnloadActivity.scala]:
````
seekEoe(next, quote.forAll(_ == quoteChar), expPart + curChar)
````

File: contrib/activity/definition/src/main/scala/com/krux/hyperion/activity/RedshiftUnloadActivity.scala (hunk 5 above)
Comment (realstraw) [inline contrib/activity/definition/src/main/scala/com/krux/hyperion/activity/RedshiftUnloadActivity.scala]: Good idea! Will change.

File: contrib/activity/definition/src/main/scala/com/krux/hyperion/activity/RedshiftUnloadActivity.scala [hunk 7]
```diff
@@ -37,13 +40,86 @@ case class RedshiftUnloadActivity private (
...
+      }
+    }
+  }
+
+  @tailrec
+  private def prepareScript(
+      exp: String, hashSpotted: Boolean = false, result: StringBuilder = StringBuilder.newBuilder
+    ): String = {
+
+    if (exp == "") {
```
Comment (sethyates) [inline contrib/activity/definition/src/main/scala/com/krux/hyperion/activity/RedshiftUnloadActivity.scala]: `exp.isEmpty`

File: contrib/activity/definition/src/main/scala/com/krux/hyperion/activity/RedshiftUnloadActivity.scala [hunk 8]
```diff
@@ -37,13 +40,86 @@ case class RedshiftUnloadActivity private (
...
+      val curChar = exp.head
+      val expTail = exp.tail
+
+      if (!hashSpotted) {  // outside a expression block
+        if (curChar == '#')
+          prepareScript(expTail, true, result + curChar)
+        else
+          prepareScript(
+            expTail,
+            hashSpotted,
```
Comment (sethyates) [inline contrib/activity/definition/src/main/scala/com/krux/hyperion/activity/RedshiftUnloadActivity.scala]: Or `false`

File: contrib/activity/definition/src/main/scala/com/krux/hyperion/activity/RedshiftUnloadActivity.scala [hunk 9]
```diff
@@ -37,13 +40,86 @@ case class RedshiftUnloadActivity private (
...
+            if (curChar == '\'') result ++= "\\\\'" else result + curChar
```
Comment (sethyates) [inline contrib/activity/definition/src/main/scala/com/krux/hyperion/activity/RedshiftUnloadActivity.scala]: What's the implication of doing `result ++= x` vs `result + y`?

File: contrib/activity/definition/src/main/scala/com/krux/hyperion/activity/RedshiftUnloadActivity.scala [hunk 10]
```diff
@@ -37,13 +40,86 @@ case class RedshiftUnloadActivity private (
...
+          )
+      } else {  // the previous char is '#'
+        if (curChar == '{') {  // start of an expression
+          val (blockBody, rest) = seekEoe(expTail)
+          prepareScript(rest, hashSpotted, result + curChar ++= blockBody)
```
Comment (sethyates) [inline contrib/activity/definition/src/main/scala/com/krux/hyperion/activity/RedshiftUnloadActivity.scala]: Or `hashSpotted` = `true`

Comment (sethyates) [inline contrib/activity/definition/src/main/scala/com/krux/hyperion/activity/RedshiftUnloadActivity.scala]: And I don't get what's happening here: `result + curChar ++= blockBody`.

File: contrib/activity/definition/src/main/scala/com/krux/hyperion/activity/RedshiftUnloadActivity.scala [hunk 11]
```diff
@@ -37,13 +40,86 @@ case class RedshiftUnloadActivity private (
...
+        } else {  // not start of an expression
+          prepareScript(
+            expTail,
+            hashSpotted,
+            if (curChar == '\'') result ++= "\\\\'" else result + curChar
```
Comment (sethyates) [inline contrib/activity/definition/src/main/scala/com/krux/hyperion/activity/RedshiftUnloadActivity.scala]: Same here with the `++=` - is there a clearer way to do this?

File: contrib/activity/definition/src/main/scala/com/krux/hyperion/activity/RedshiftUnloadActivity.scala (hunk 9 above)
Comment (realstraw) [inline contrib/activity/definition/src/main/scala/com/krux/hyperion/activity/RedshiftUnloadActivity.scala]: `++=` takes `String`, `+` takes `Char`

File: contrib/activity/definition/src/main/scala/com/krux/hyperion/activity/RedshiftUnloadActivity.scala (hunk 10 above)
Comment (realstraw) [inline contrib/activity/definition/src/main/scala/com/krux/hyperion/activity/RedshiftUnloadActivity.scala]: append `curChar: String` and then append `blockBody: String`

File: contrib/activity/definition/src/main/scala/com/krux/hyperion/activity/RedshiftUnloadActivity.scala (hunk 9 above)
Comment (sethyates) [inline contrib/activity/definition/src/main/scala/com/krux/hyperion/activity/RedshiftUnloadActivity.scala]: I see.  I was thinking this whole thing might be clearer (but maybe not as performant?) if you converted it to a `List` to begin with and pattern matched on the list syntax.

File: contrib/activity/definition/src/main/scala/com/krux/hyperion/activity/RedshiftUnloadActivity.scala (hunk 11 above)
Comment (realstraw) [inline contrib/activity/definition/src/main/scala/com/krux/hyperion/activity/RedshiftUnloadActivity.scala]: I guess I should change `+` to `+=` to make it consistent.

File: contrib/activity/definition/src/main/scala/com/krux/hyperion/activity/RedshiftUnloadActivity.scala (hunk 9 above)
Comment (realstraw) [inline contrib/activity/definition/src/main/scala/com/krux/hyperion/activity/RedshiftUnloadActivity.scala]: Changing `+` to `+=` should make it clearer and consistent. The problem with List[Char] is that it's a bit hard to convert it to String, probably have to do something like `theList.mkString("")`

File: contrib/activity/definition/src/main/scala/com/krux/hyperion/activity/RedshiftUnloadActivity.scala (hunk 11 above)
Comment (sethyates) [inline contrib/activity/definition/src/main/scala/com/krux/hyperion/activity/RedshiftUnloadActivity.scala]: Would be clearer what's going on.

Comment (sethyates) [inline contrib/activity/definition/src/main/scala/com/krux/hyperion/activity/RedshiftUnloadActivity.scala]: Still looks odd doing a `?=` in the method call here.

File: contrib/activity/definition/src/main/scala/com/krux/hyperion/activity/RedshiftUnloadActivity.scala (hunk 9 above)
Comment (sethyates) [inline contrib/activity/definition/src/main/scala/com/krux/hyperion/activity/RedshiftUnloadActivity.scala]: OK.  Having a consistent `=` in there would help.

File: contrib/activity/definition/src/main/scala/com/krux/hyperion/activity/RedshiftUnloadActivity.scala (hunk 10 above)
Comment (sethyates) [inline contrib/activity/definition/src/main/scala/com/krux/hyperion/activity/RedshiftUnloadActivity.scala]: How about `prepareScript(rest, hashSpotted, result ++= s"$curChar$blockBody")`

File: contrib/activity/definition/src/main/scala/com/krux/hyperion/activity/RedshiftUnloadActivity.scala (hunk 11 above)
Comment (sethyates) [inline contrib/activity/definition/src/main/scala/com/krux/hyperion/activity/RedshiftUnloadActivity.scala]: Perhaps `result ++= quoteChar(curChar)` where `quoteChar` is:
````
def quoteChar(curChar: Char): String = if (curChar == '\'') "\\\\'" else curChar.toString
````

File: contrib/activity/definition/src/main/scala/com/krux/hyperion/activity/RedshiftUnloadActivity.scala [hunk 12]
```diff
@@ -37,13 +40,86 @@ case class RedshiftUnloadActivity private (
...
+
+    if (exp == "") {
+      result.toString
+    } else {
+      val curChar = exp.head
+      val expTail = exp.tail
+
+      if (!hashSpotted) {  // outside a expression block
+        if (curChar == '#')
+          prepareScript(expTail, true, result + curChar)
```
Comment (sethyates) [inline contrib/activity/definition/src/main/scala/com/krux/hyperion/activity/RedshiftUnloadActivity.scala]: Same here with `result += curChar`?

File: contrib/activity/definition/src/main/scala/com/krux/hyperion/activity/RedshiftUnloadActivity.scala (hunk 11 above)
Comment (realstraw) [inline contrib/activity/definition/src/main/scala/com/krux/hyperion/activity/RedshiftUnloadActivity.scala]: Sounds good, I'll change to use `def` instead

File: contrib/activity/definition/src/main/scala/com/krux/hyperion/activity/RedshiftUnloadActivity.scala [hunk 13]
```diff
@@ -37,13 +40,77 @@ case class RedshiftUnloadActivity private (
...
+
+    if (exp.isEmpty) {
+      throw new RuntimeException("expression started but not ended")
+    } else {
+      val curChar = exp.head
+      val next = exp.tail
+
+      quote match {
+        case Some(quoteChar) =>  // if is in quote
+          if (curChar == quoteChar) seekEndOfExpr(next, None, expPart += curChar)
```
Comment (sethyates) [inline contrib/activity/definition/src/main/scala/com/krux/hyperion/activity/RedshiftUnloadActivity.scala]: You didn't like `quote.filter(_ == curChar)`? ;-)

File: contrib/activity/definition/src/main/scala/com/krux/hyperion/activity/RedshiftUnloadActivity.scala [hunk 14]
```diff
@@ -37,13 +40,77 @@ case class RedshiftUnloadActivity private (
...
+          else seekEndOfExpr(next, quote, expPart += curChar)
+        case _ =>
+          if (curChar == '}')
```
Comment (sethyates) [inline contrib/activity/definition/src/main/scala/com/krux/hyperion/activity/RedshiftUnloadActivity.scala]:
````
curChar match {
case '}' => ((expPart += curChar).result, next)
case '\'' | '"' => seekEndOfExpr(next, Option(curChar), expPart += curChar)
case _ => seekEndOfExpr(next, None, expPart += curChar)
}
````

File: contrib/activity/definition/src/main/scala/com/krux/hyperion/activity/RedshiftUnloadActivity.scala [hunk 15]
```diff
@@ -37,13 +40,77 @@ case class RedshiftUnloadActivity private (
...
+            seekEndOfExpr(next, Option(curChar), expPart += curChar)
+          else
+            seekEndOfExpr(next, None, expPart += curChar)
+      }
+    }
+  }
+
+  @tailrec
+  private def prepareScript(
+      exp: String, hashSpotted: Boolean = false, result: StringBuilder = StringBuilder.newBuilder
```
Comment (sethyates) [inline contrib/activity/definition/src/main/scala/com/krux/hyperion/activity/RedshiftUnloadActivity.scala]: One per line if using this format.

File: contrib/activity/definition/src/main/scala/com/krux/hyperion/activity/RedshiftUnloadActivity.scala [hunk 16]
```diff
@@ -37,13 +40,77 @@ case class RedshiftUnloadActivity private (
...
+    ): String = {
+
+    if (exp.isEmpty) {
+      result.toString
+    } else {
+      val curChar = exp.head
+      val expTail = exp.tail
+
+      def appendEscapedChar: StringBuilder = if (curChar == '\'') result ++= "\\\\'" else result += curChar
```
Comment (sethyates) [inline contrib/activity/definition/src/main/scala/com/krux/hyperion/activity/RedshiftUnloadActivity.scala]: Not quite what I had in mind.  This style still mixes usage as below and also closes over several things.  I had in mind:
````
def quoteChar(ch: Char): String = if (ch == '\'') "\\\\'" else curChar.toString
````

then the usage would be below as:
````
result ++= escapeChar(curChar)
````

File: contrib/activity/definition/src/main/scala/com/krux/hyperion/activity/RedshiftUnloadActivity.scala [hunk 17]
```diff
@@ -37,13 +40,77 @@ case class RedshiftUnloadActivity private (
...
+
+      if (!hashSpotted) {  // outside a expression block
+        if (curChar == '#')
```
Comment (sethyates) [inline contrib/activity/definition/src/main/scala/com/krux/hyperion/activity/RedshiftUnloadActivity.scala]: You could collapse this whole block to:
````
prepareScript(expTail, curChar == '#', result ++= quoteChar(curChar))
````

**Comment by sethyates** [comment]
LGTM.  This should work.  Remaining comments are purely stylistic.

File: contrib/activity/definition/src/main/scala/com/krux/hyperion/activity/RedshiftUnloadActivity.scala (hunk 13 above)
Comment (realstraw) [inline contrib/activity/definition/src/main/scala/com/krux/hyperion/activity/RedshiftUnloadActivity.scala]: I guess you referring to replace `if else` with the `filter`? But, it's actually different. `quote.filter(_ == curChar)` will change quote to None in the `else` case.

Comment (sethyates) [inline contrib/activity/definition/src/main/scala/com/krux/hyperion/activity/RedshiftUnloadActivity.scala]: Sorry I mean `!=`

Comment (realstraw) [inline contrib/activity/definition/src/main/scala/com/krux/hyperion/activity/RedshiftUnloadActivity.scala]: But `!=` will not change it to `None` in the `if` case.

Comment (sethyates) [inline contrib/activity/definition/src/main/scala/com/krux/hyperion/activity/RedshiftUnloadActivity.scala]: Huh?
````
scala> val quote = Some('"')
quote: Some[Char] = Some(")
scala> val curChar = '"'
curChar: Char = "
scala> quote.filter(_ != curChar)
res1: Option[Char] = None
````

````
scala> val quote = Some('"')
quote: Some[Char] = Some(")
scala> val curChar = 'x'
curChar: Char = x
scala> quote.filter(_ != curChar)
res3: Option[Char] = Some(")
````

Comment (realstraw) [inline contrib/activity/definition/src/main/scala/com/krux/hyperion/activity/RedshiftUnloadActivity.scala]: You're right, not sure what I was thinking back then :-)

**Comment by sethyates** [comment]
:+1: nailed it :sparkles:  :sparkles:  :sparkles:

**Pull request closed without merging**